### ISSUE TITLE
`HTTP::Client` mutates request headers

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#12548]
## Description

When sending an HTTP request through `HTTP::Client`, the client sends some implicit headers with the request. The `Host` and `User-Agent` headers apply on every request.
Further, `Accept-Encoding` , `Connection`, `Authorization`, `Content-Type`, `Transfer-Encoding`, `Content-Length` and `Cookie` depending on conditions.

All these headers are added to the request headers while the request is handled in the client, thus mutating the request.

```cr
require "http/client"

client = HTTP::Client.new(IO::Memory.new)
request = HTTP::Request.new("GET", "/")
request.headers # => HTTP::Headers{}
client.exec(request) rescue nil # Ignore defunct connection, we're only interested in the side effects on `request`
request.headers # => HTTP::Headers{"Host" => "", "User-Agent" => "Crystal", "Accept-Encoding" => "gzip, deflate"}
```

### Shit hits the fan...

If that same request instance was re-used to make another HTTP request, it would now have a different effect. Particularly relevant is the addition of the `Accept-Encoding` header which instructs the client *not* to implicitly decompress the body.

```cr
require "http/client"

client = HTTP::Client.new("crystal-lang.org", tls: true)
request = HTTP::Request.new("GET", "/")
response = client.exec(request)
# response is sent compressed, and HTTP::Client applies implicit decompression
response.body[..10] # => "<!DOCTYPE h"
response = client.exec(request)
# response is sent compressed, but no implicit decompression because of the explicit Accept-Encoding header
response.body[..10] # => "\u001F\x8B\b\u0000\u0000\u0000\u0000\u0000\u0000\u0003|"
```

This is basically the same problem as #11354, just spread over separate request cycles. #12536 fixed the issue for the internal retry mechanism. But this problem still exists when re-using a request instance.

## Discussion

While the effects for other headers may not be as catastrophic, I think it's a general problem. Executing a request should not have side effects on the request object. When re-using it to execute again, I would expect *exactly* the same behaviour and not observe some alterations introduced by the first execution.

All these additional headers have only internal purposes. There's no external reason to add them to the request object.[^1] It's just a convenient solution to mutate the request.
But we can change the implementation to do without that. It's just an internal detail, so this should have limited external effects - except for no longer mutating the request headers of course.

## Proposal

For that we need to convey the additional header information alongside the request object (instead of injecting it into it) so it can be used when serializing the headers.
I have a working implementation for this in https://github.com/crystal-lang/crystal/compare/master...straight-shoota:fix/http-client-request-mutation
The data is stored in a `TransmissionRecord`[^2] struct which is passed from the client to the header serialization method.

Note that it's not complete (yet):
* `Cookie` is still added to the request headers. This can be modeled into the `TransmissionRecord`.
* `Content-Type` headers (for form submissions) are still added to the request. But I think this might actually be fine; it's only set when explicitly sending a request with form content.
* The handling of `Transfer-Encoding` is perhaps incorrect (it will be ignored as a request header), but I'm not sure about actual use cases.

[^1]: One might argue that it allows to inspect the *actual* header data that has been sent. I think that's a weak argument, though. This behaviour is very unexpected. If inspection of exact message data is of relevance, we can add a logging mechanism for that.
[^2]: Suggestions for a better name are welcome, btw.

**Comment by asterite** [comment]
I think the solution is to document that you shouldn't reuse a Request object.

**Comment by straight-shoota** [comment]
> I think in Go the request is also mutated.

No, it's not. Go actually uses a similar implementation as the one I have linked. It's just a bit more complex due to several layers of indirection. See for example https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/net/http/request.go;l=629;drc=61ed6d5c3341f73af9529b4808dd0997c6c86ed4
